Add rst-reader back again

raskchanky opened the pull request:
This reverts https://github.com/habitat-sh/habitat/pull/6616 which reverted the addition of the `rst-reader` tool.  In addition to that reversion, it also accounts for the case where a supervisor is started up fresh for the first time, and no rumor file exists yet.  Previously, the supervisor would panic, which was sub-optimal.  Now we just create an empty file if we need to.

Resolves https://github.com/habitat-sh/habitat/issues/3169

![](https://media.giphy.com/media/xUA7bb7owFcrNPAw7u/giphy.gif)

mwrock commented:
I'm wondering if you rebased if the clippy errors go away.

raskchanky commented:
@mwrock It's already been rebased against master.  I tried again just now and it's up to date.  I think the rust 1.35.0 PR has some fixes as well that will help.

baumanj commented on components/butterfly/src/rumor/dat_file.rs:
```diff
@@ -43,7 +44,11 @@ pub struct DatFile {
+                                     .read(true)
+                                     .write(true)
+                                     .open(&data_path)
+                                     .map_err(|err| Error::DatFileIO(data_path.clone(), err))?;
```
I think this is going to potentially break the atomicity of writing this file. Especially now that we potentially have multiple processes accessing it, it's important that we access this safely.

I think we could re-work this a bit more such that
1. If the supervisor calls this and there's no existing .rst file, it's created via the standard `DatFile::write` method, which will use `AtomicWriter`
1. If `rst-reader` calls this and there's no existing .rst file, it simply returns `None`/`Err`; I think an empty file vs no file is an important distinction.

Probably splitting this into two functions `DatFile::read_or_create(data_path: PathBuf, server: &Server) -> Result<Self>` and `DatFile::read() -> Option<Self>` for those two cases would make it more ergonomic. This would also eliminate the need for the `if size == 0` check in `read_into`.

I'm also a little concerned that the the abstraction here is still a bit off since we have calls like
```rust
dat_file.read_rumors::<Service>(dat_file.service_len())
```
Consumers needing to tell `dat_file` exactly the offset to read at seems like a fairly big hole in the encapsulation. This is a problem with the initial implementation, not your additions. I think it would be good to keep re-working it a bit more so we can get at something more usable.

raskchanky replied:
Previous to the addition of `rst-reader`, the only place that `DatFile::new()` was called was within the butterfly server when it started up.  Would giving `rst-reader` its own method to use, e.g. `DatFile::read()` as you suggested, solve the concerns around creating the file atomically?  I realize that doesn't provide any type of atomicity, per se, but it does resolve the issue of multiple processes trying to create the file.

I bring it up because `DatFile::write` takes `self`, which implies a certain chicken/egg problem when using it inside an initializer.  I tried changing `DatFile::write` to be a static method, but then the call site in the butterfly server ends up looking like `DatFile::write(dat_file.path(), self)`, which I dunno...feels a little odd.

Is there an ergonomic way to make this work that I'm missing?

raskchanky replied:
Nevermind, I think I have a solution to this.  Also, your point about consumers needing to tell `dat_file` the offset is taken.  That also felt off to me.  I'll update this to account for that.

raskchanky commented on components/butterfly/src/rumor/dat_file.rs:
```diff
@@ -46,163 +39,171 @@ pub struct DatFile {
+                              reader }))
+        } else {
+            None
+        }
```
Nice, that's much better.

raskchanky commented on components/rst-reader/src/main.rs:
```diff
@@ -0,0 +1,143 @@
+            error!("Dat file at {} does not exist.", path.display());
+            process::exit(1);
+        }
+    };
```
Excellent

raskchanky commented on components/butterfly/src/server/mod.rs:
```diff
@@ -416,7 +416,10 @@ impl Server {
+            let dat_path = path.join(format!("{}.rst", &self.member_id));
+            let mut file = DatFile::read_or_create(dat_path, &self)?;
+
             if file.path().exists() {
```
Ah, good call.

raskchanky commented:
@baumanj I believe I've addressed all your feedback.  I also rebased this onto master, and there were a number of conflicts with the work you did in https://github.com/habitat-sh/habitat/pull/6662.  I believe I've sorted those out correctly (everything compiles and appears to work) but would appreciate another look through this when you get the chance.

baumanj commented on components/butterfly/src/rumor/dat_file.rs:
```diff
@@ -46,166 +56,156 @@ pub struct DatFile {
-        let mut reader = BufReader::new(file);
-        reader.read_exact(&mut version)
-              .map_err(|err| Error::DatFileIO(self.path.clone(), err))?;
+    pub fn read_header(&mut self, version: &mut [u8]) -> Result<()> {
```
Requiring the caller to know the internal details, like the size of the header seems very error prone. Instead, why not make `read_header` a private method which is always called when a `DatFile` value is created?

Additionally, there's no need for the consumers to know the version. If we change `offset_for_rumor` to return an `Option<64>` rather than panicking when `message_id` is an rumor type, then callers don't have any need for it.

raskchanky replied:
Yeah good call

baumanj commented on components/butterfly/src/rumor/dat_file.rs:
```diff
@@ -358,42 +341,62 @@ impl DatFile {
-            1 => vec![0; mem::size_of::<Self>()],
-            _ => vec![0; mem::size_of::<Self>() + 8],
+            1 => vec![0; HEADER_VERSION_1_SIZE],
+            _ => vec![0; HEADER_VERSION_2_SIZE],
```
This should probably be
```suggestion
            2 => vec![0; HEADER_VERSION_2_SIZE],
            _ => unimplemented!(),
```
Or perhaps something a bit more user-friendly.

raskchanky replied:
Ah, yes, good thinking.

raskchanky commented on components/butterfly/src/rumor/dat_file.rs:
```diff
@@ -29,9 +13,35 @@ use crate::{error::{Error,
+// Header to hold a HashMap of MESSAGE_ID -> offset, we can't rely on std::mem::size_of to give us
+// the correct information any more.
+const HEADER_VERSION_1_SIZE: usize = 48;
+const HEADER_VERSION_2_SIZE: usize = 64;
```
The math is slightly off here, I think, since version 1 had 6 fields and version 2 had 7 fields, but I get your point.  I can rework this.

raskchanky commented:
It's worth noting that in the course of addressing PR feedback, I uncovered an issue where previously, if butterfly tried to start up and the rumor file was corrupt, it would silently succeed.  With the changes I've made, that's no longer the case.  Now if `DatFile` detects that the file is corrupt, it will abort with an error.
